src/DataContainers.jl

```diff
@@ -16,9 +16,9 @@ Container to store data samples as columns in an array.
 """
 struct DataContainer{FT <: Real}
     #stored data, each piece of data is a column [data dimension × number samples]
-    stored_data::Array{FT, 2}
+    stored_data::AbstractMatrix{FT}
```

Comment:
https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-fields-with-abstract-containers

Reply:
I would say most of infa here isn't performance critical

Reply:
A concern that can be limiting for the package is unnecessary memory use. In some cases, we want to store covariances that are extremely sparse, or even diagonal. Casting them and storing them as a Matrix{FT} then takes up much more space than e.g. Diagonal {FT}. Whichever solution solves this problem would work. Let me know how to proceed.

Reply:
I would say write it as generically as possible and then if there is some performance hotspots after a bit of profiling they should be straightforward to address.  You don't need to specialize on every Matrix type here so IMO what is written is fine.

Reply:
I wasn't suggesting to specialize on every type of matrix, just to make the type (and its properties) concrete. This was @glwagner's [suggestion](https://github.com/CliMA/EnsembleKalmanProcesses.jl/issues/99#issuecomment-1022773454), too. It's not much overhead to have very concrete types and simply not specialize on methods that use it.

Reply:
then you cannot mix and mach representations (different matrix types might be required for different fields).  Until performance is actually an issue I wouldn't worry about it

Reply:
for calibration the costs are going to be dominated by actually running the models and IO, everything else I'm guessing won't matter much but we'll see as we go

Reply:
> A concern that can be limiting for the package is unnecessary memory use. In some cases, we want to store covariances that are extremely sparse, or even diagonal. Casting them and storing them as a Matrix{FT} then takes up much more space than e.g. Diagonal {FT}. Whichever solution solves this problem would work. Let me know how to proceed.

@ilopezgp the point that @charleskawczynski was making is the difference between

```julia
struct DataContainer{FT <: Real}
    #stored data, each piece of data is a column [data dimension × number samples]
    stored_data::Array{FT, 2}
    stored_data::AbstractMatrix{FT}
```

and 

```julia
struct DataContainer{M <: AbstractMatrix}
    #stored data, each piece of data is a column [data dimension × number samples]
    stored_data::M
```

In the first case, `stored_data` is _abstractly typed_, which means that the compiler cannot infer the type of `stored_data` given the type `DataContainer`. This has performance penalties (which @jakebolewski argues are not important). But it's utterly fundamental and so an important point to grasp.

Writing `M <: AbstractMatrix` will preclude `UniformScaling`. Omitting `<: AbstractMatrix` is the conservative choice. You might also write `M <: Union{AbstractMatrix, UniformScaling}`. However this is risky because you may be failing to anticipate other valid matrix-like types.

Reply:
I understand the difference between abstract types and parameterized types, thank you. I think the conversation was about whether it practically matters in this application.